xonsh/aliases.py

```diff
@@ -94,6 +95,10 @@ def source_foreign(args, stdin=None):
         if k in denv and v == denv[k]:
             continue  # no change from original
         env[k] = v
+    # If run in un-safe mode we are sure the command completed correctly,
+    # thus we can remove any env-vars that were unset by the script.
+    if not ns.safe:
+        [env.pop(k, None) for k in denv if k not in fsenv]
```

Comment:
I added some functionality to let the sources scripts also unset the environments variables. We just need to be sure that `fsenv` not empty because it failed to run.